js/lib/menuUtil.js

```diff
@@ -0,0 +1,166 @@
+    }
+  }
+  return null
+}
+
+/**
+ * Check for uneeded updates.
+ * Updating the menu when it is not needed causes the menu to close if expanded
+ * and also causes menu clicks to not work.  So we don't want to update it a lot.
+ * Should only be updated when appState or windowState change (for history or bookmarks)
+ * NOTE: settingsState is not used directly; it gets used indirectly via getSetting()
+ * @param {Object} appState - Application state
+ * @param {Object} windowState - Current window state
+ */
+module.exports.checkForUpdate = (appState, windowState) => {
+  const updated = {
+    nothing: true,
+    settings: false,
+    sites: false,
+    closedFrames: false
+  }
+
+  if (appState && appState.get('settings') !== lastSettingsState) {
+    // Currently only used for the HOMEPAGE value (bound to history menu)
+    lastSettingsState = appState.get('settings')
+    updated.nothing = false
+    updated.settings = true
+  }
+
+  if (appState && appState.get('sites') !== lastSites) {
+    lastSites = appState.get('sites')
+    updated.nothing = false
+    updated.sites = true
+  }
+
+  if (windowState && windowState.closedFrames !== lastClosedFrames) {
+    lastClosedFrames = windowState.closedFrames
+    updated.nothing = false
+    updated.closedFrames = true
+  }
+
+  return updated
+}
+
+const createBookmarkMenuItems = (bookmarks, parentFolderId) => {
+  const filteredBookmarks = parentFolderId
+    ? bookmarks.filter((bookmark) => bookmark.get('parentFolderId') === parentFolderId)
+    : bookmarks.filter((bookmark) => !bookmark.get('parentFolderId'))
+
+  const payload = []
+  filteredBookmarks.forEach((site) => {
+    if (site.get('tags').includes(siteTags.BOOKMARK) && site.get('location')) {
+      payload.push({
+        // TODO include label made from favicon. It needs to be of type NativeImage
+        // which can be made using a Buffer / DataURL / local image
+        // the image will likely need to be included in the site data
+        // there was potentially concern about the size of the app state
+        // and as such there may need to be another mechanism or cache
+        //
+        // see: https://github.com/brave/browser-laptop/issues/3050
+        label: site.get('customTitle') || site.get('title'),
+        click: (item, focusedWindow, e) => {
+          if (eventUtil.isForSecondaryAction(e)) {
+            CommonMenu.sendToFocusedWindow(focusedWindow, [messages.SHORTCUT_NEW_FRAME, site.get('location'), { openInForeground: !!e.shiftKey }])
+          } else {
+            CommonMenu.sendToFocusedWindow(focusedWindow, [messages.SHORTCUT_ACTIVE_FRAME_LOAD_URL, site.get('location')])
+          }
+        }
+      })
+    } else if (siteUtil.isFolder(site)) {
+      const folderId = site.get('folderId')
+      const submenuItems = bookmarks.filter((bookmark) => bookmark.get('parentFolderId') === folderId)
+      payload.push({
+        label: site.get('customTitle') || site.get('title'),
```

Comment:
This one is a little weird, since it's a folder (and doesn't have a location). But I did implement location for the one above :smile: